### PR TITLE
add an admission decorator that allows skipping some admission checks…

### PR DESCRIPTION
**Pull request by deads2k**
… based on namespace labels

Needed to split apiservers.  This provides a decorator that optionally wraps an admission plugin in a "skip if namespace is labelled a certain way" function.  `openshift.io/run-level=1` is the intent.

/assign @sttts 
/assign @mfojtik 

@openshift/sig-master

### CODE REVIEW

File: pkg/admission/namespaceconditions/decorator.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/admission"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1lister "k8s.io/client-go/listers/core/v1"
+)
+
+// NamespaceLabelConditions provides a decorator that can delegate and conditionally add label conditions
+type NamespaceLabelConditions struct {
+	Delegate admission.Decorator
```
Comment (sttts) [inline pkg/admission/namespaceconditions/decorator.go]: concatenation of decorators looks orthogonal. Why this delegate?

I would expect `admission.Concat(name NamespaceLabelConditions.WithNamespaceLabelConditions, someOtherDecorator, yetAnotherDecorator)` if you really want to avoid passing `name` multiple time in a manual concatenation.

Comment (deads2k) [reply pkg/admission/namespaceconditions/decorator.go]: > concatenation of decorators looks orthogonal. Why this delegate?
> 
> I would expect admission.Concat(name NamespaceLabelConditions.WithNamespaceLabelConditions, someOtherDecorator, yetAnotherDecorator) if you really want to avoid passing name multiple time in a manual concatenation.

Yeah, I can make a decorator chain upstream for us to get in 1.10.  Functionally acceptable otherwise?

Comment (sttts) [reply pkg/admission/namespaceconditions/decorator.go]: Yes, am a big fan of decorators. And we obviously need this.

File: pkg/admission/namespaceconditions/labelcondition.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apiserver/pkg/admission"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1lister "k8s.io/client-go/listers/core/v1"
+)
+
+const runLevelLabel = "openshift.io/run-level"
```
Comment (mfojtik) [inline pkg/admission/namespaceconditions/labelcondition.go]: runlevel?

Comment (mfojtik) [reply pkg/admission/namespaceconditions/labelcondition.go]: maybe something like `openshift.io/infra-runlevel`

File: pkg/cmd/server/origin/admission/chain_builder.go [hunk 3]
```diff
@@ -30,6 +29,22 @@ import (
 )
 
 var (
+	// these are admission plugins that cannot be applied until after the kubeapiserver starts.
+	// TODO if nothing comes to mind in 3.10, kill this
+	SkipRunLevelZeroPlugins = sets.NewString()
+	// these are admission plugins that cannot be applied until after the openshiftapiserver apiserver starts.
+	SkipRunLevelOnePlugins = sets.NewString(
+		"ProjectRequestLimit",
```
Comment (mfojtik) [inline pkg/cmd/server/origin/admission/chain_builder.go]: shouldn't we make this configurable externally? (master config?)

Comment (deads2k) [reply pkg/cmd/server/origin/admission/chain_builder.go]: > shouldn't we make this configurable externally? (master config?)

I don't think so.  We know which ones require openshift resources.  We aren't providing generic skipping possbilities.

File: pkg/admission/namespaceconditions/labelcondition.go (hunk 2 above)
Comment (deads2k) [reply pkg/admission/namespaceconditions/labelcondition.go]: runlevel matches upstream docs and gives clear cardinality.  https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/admissionregistration/v1beta1/types.go#L152
